pkg/capabilities/consensus/ocr3/datafeeds/llo_aggregator.go

```
@@ -0,0 +1,409 @@
+		timeDiffNs := observationTimestamp - uint64(previousStreamInfo.Timestamp) //nolint:gosec // G115
+		lggr.Debugw("checking deviation and heartbeat",
+			"streamID", streamID,
+			"observationNs", observationTimestamp,
+			"perviousNs", previousStreamInfo.Timestamp,
+			"currStalenessNs", timeDiffNs,
+			"heartbeatSec", config.Heartbeat,
+			"oldPrice", oldPrice,
+			"newPrice", newPrice,
+			"currDeviation", priceDeviation,
+			"deviation", config.Deviation.InexactFloat64(),
+		)
+		if timeDiffNs > uint64(config.HeartbeatNanos()) || //nolint:gosec // G115
+			priceDeviation > config.Deviation.InexactFloat64() {
+			// this stream needs an update
+			previousStreamInfo.Timestamp = int64(observationTimestamp) //nolint: gosec // G115
+			var err2 error
+			previousStreamInfo.Price, err2 = prices[streamID].MarshalBinary()
+			if err2 != nil {
+				lggr.Errorw("failed to marshal price", "streamID", streamID, "err", err2)
+				continue
+			}
+			mustUpdateIDs = append(mustUpdateIDs, streamID)
+		} else if float64(timeDiffNs) > float64(config.HeartbeatNanos())*(1.0-a.config.allowedPartialStaleness) {
+			maybeUpdateIDs = append(maybeUpdateIDs, streamID)
+		}
+	}
+
+	// optimization that allows for more efficient batching
+	// if there is at least one stream that actually hit its deviation or heartbeat threshold,
+	// append all others that were within AllowedPartialStaleness percentage of their heartbeat
+	if len(mustUpdateIDs) > 0 {
+		mustUpdateIDs = append(mustUpdateIDs, maybeUpdateIDs...)
+		// deterministic order
+		slices.Sort(mustUpdateIDs)
+	}
+
+	marshalledState, err := proto.MarshalOptions{Deterministic: true}.Marshal(currentState)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal current state: %w", err)
+	}
+
+	toWrap := make([]*WrappableStreamUpdate, 0, len(mustUpdateIDs))
+	for _, streamID := range mustUpdateIDs {
+		// TODO what if remapped ID is not defined? How do we reconcile binary vs int? Should remapped IDs also be integers now?
```

Maybe return an error if remapped ID is not defined, at least until we resolve it?

---

made remappedIDHex required in the config parser for llo streams